tidy3d/plugins/design/method.py

```diff
@@ -180,23 +124,430 @@ def sample(self, parameters: Tuple[ParameterType, ...], **kwargs) -> Dict[str, A
+        ...,
+        title="Number of solutions per population.",
+        description="TBD",
+    )
+
+    n_generations: pd.PositiveInt = pd.Field(
+        ...,
+        title="Number of generations.",
+        description="TBD",
+    )
+
+    n_parents_mating: pd.PositiveInt = pd.Field(
+        ...,
+        title="Number of parents mating.",
+        description="TBD",
+    )
+
+    stop_criteria: Optional[pd.constr(regex=r"\b(?:reach|saturate)_\d+\b")] = pd.Field(
+        title="Early stopping criteria.",
+        description="Define the early stopping criteria. Supported words are 'reach_X' or 'saturate_X' where X is a number. See PyGAD docs for more details",
+        default=None,
+    )
+
+    parent_selection_type: Optional[
+        Literal["sss", "rws", "sus", "rank", "random", "tournament"]
+    ] = pd.Field(
+        title="Parent selection type.",
+        description="TBD",
+        default="sss",
+    )
+
+    crossover_type: Optional[Literal["single_point", "two_points", "uniform", "scattered"]] = (
+        pd.Field(
+            title="Crossover type.",
+            description="TBD",
+            default="single_point",
+        )
+    )
+
+    crossover_prob: Optional[pd.confloat(ge=0, le=1)] = pd.Field(
+        title="Crossover probability.",
+        description="TBD",
+        default=0.8,
+    )
+
+    mutation_type: Optional[Literal["random", "swap", "inversion", "scramble", "adaptive"]] = (
+        pd.Field(
+            title="Crossover type.",
+            description="TBD",
+            default="random",
+        )
+    )
+
+    mutation_prob: Optional[pd.confloat(ge=0, le=1)] = pd.Field(
+        title="Crossover probability.",
+        description="TBD",
+        default=0.2,
+    )
+
+    # TODO: See if anyone is interested in having the full suite of PyGAD options - there's a lot!
+
+    def run(self, parameters: Tuple[ParameterType, ...], run_fn: Callable) -> Tuple[Any]:
+        """Defines the search algorithm for the GA"""
+
+        # Create fitness function combining pre and post fn with the tidy3d call
+        def fitness_function(ga_instance, solution, solution_idx):
+            # Break solution down to dict
+            sol_dict = self.sol_array_to_dict(solution, _param_keys, _param_converter)
+
+            # Iterate through solutions for batched and non-batched data
+            sol_out, aux_out = self._extract_output(run_fn(sol_dict))
+
+            self._flatten_and_append(aux_out, _store_aux)
+
+            return sol_out
+
+        def on_generation(ga_instance):
+            _store_parameters.append(ga_instance.population.copy())
+            _store_fitness.append(ga_instance.last_generation_fitness)
+            best_fitness = ga_instance.best_solution()[1]
+            print(
+                f"Generation {ga_instance.generations_completed}: Best Fitness = {best_fitness:.3f}"
+            )
+
+        # Make param names available to the fitness function
+        global _param_keys
```

Comment:
We should avoid this if at all possible.  Can we just pass these to our fitness function as kwargs?

Reply:
Unfortunately not, neither PyGAD or PySwarms allow their fitness function definitions in include any other arguments except for the ones it expects. A lot of their examples show them using globals or just defining variables in the same namespace as the function.

Reply:
Hm, so it doesn't work if you define `param_keys` before the fitness function?
```
_param_keys = [param.name for param in parameters]

def fitness_function(...)
    # use _param_keys_
```